*20120172 - Vân Duy Quang*
## **BÀI TẬP 1: ÔN TẬP XÁC SUẤT**

#### **Câu 1.**  
Gọi $R_1$, $R_2$, $R_3$, $R_4$ lần lượt là các biến cố router $R_1$, $R_2$, $R_3$ và $R_4$ hoạt động, $E$ là biến cố $A$ liên lạc được $B$.  

#### *a)*  
Ta thấy $A$ và $B$ liên lạc được thông qua 
$
\begin{cases}
    A \rightarrow R_2 \rightarrow R_3 \rightarrow B\\
    A \rightarrow R_2 \rightarrow R_4 \rightarrow B.  
\end{cases}
$

Do đó:  
$$E_a = (R_2 \cap R_3) \cup (R_2 \cap R_4) = R_2\cap(R_3 \cup R_4)$$

Suy ra: 
\begin{align*}
    & P(E_a) &=& P(R_2.(R_3 + R_4)) \\
    &\;&=& P(R_2).P(R_3 + R_4 | R_2)\\
    &\; &=& P(R_2).[1 - P(\overline{R_3}.\overline{R_4} | R_2)]\\
    &\; &=& P(R_2).(1 - P(\overline{R_3} | R_2).P(\overline{R_4} | R_2)) \\
    &\; &=& P(R_2).(1 - P(\overline{R_3}).P(\overline{R_4})) \\
    &\; &=& 0.85 \times (1 - 0.2\times 0.25)) \\
    &\; &=& 0.8075
\end{align*}

Vậy  
$$P(E_a) = 0.8075$$


#### *b)*  
$R_1$ tạo đường đi thẳng $A \rightarrow R_1 \rightarrow B$. Vậy miễn là  $R_1$ không hư, $A$ và $B$ có thể liên lạc được, như vậy ta có:
$$E_b = R_1 \cup E_a$$

Do đó: 
$$P(E_b) = P(R_1 + E_a) = 1 - P(\overline{R_1}).P(\overline{E_a}) = 1 - P(\overline{R_1}).[1 - P(E_a)]$$

Vậy:  
$$P(E_b) = 1 - 0.1\times(1 - 0.8075) = 0.98075$$

##### **Kiểm tra**

*Giải thích:* 
* Hàm `find_path` là hàm đệ quy giúp kiểm tra xem từ vị trí  `From` có đi đến vị trí `To` được không, lấy mảng `existed_items` làm mảng chứa những phần tử đã đi qua nhằm tránh tạo chu trình đồ thị. Trước khi bắt đầu quá trình đệ quy, hàm này kiểm tra xem `From` có nằm trong danh sách `failed_items` chứa những router hỏng không, nếu router `From` hỏng, trả về `False` cho hàm cha.
* Hàm `make_fail_items` giúp tạo ra mảng `failed_items` nói trên, dựa vào thông số `fail_prob` cung cấp xác suất hỏng của từng router (Xem như $A$ và $B$ cũng là router với xác suất hỏng là $0$)

In [1]:
import random

vertexes = ['A', 'R1', 'R2', 'R3', 'R4', 'B']
edges = [('A', 'R1'), ('A', 'R2'), ('R1', 'R3'), ('R1', 'B'),
         ('R3', 'R2'), ('R3', 'B'), ('R2', 'R4'), ('R4', 'B')]
fail_prob = [0, 0.1, 0.15, 0.2, 0.25, 0]


def find_path(From = 'A', existed_items = [], To = 'B', failed_items = []):
	if (From in failed_items):
		return False
	existed_items.append(From)
	if From == To:
		# print(existed_items)
		existed_items.pop(-1)
		return True
	
	for edge in edges:
		if From in edge:
			other = edge[1] if edge[0] == From else edge[0]
			if (other in existed_items):
				continue
			if find_path(other, existed_items, To, failed_items):
				return True
	existed_items.pop(-1)
	return False

def make_fail_items():
	failed_items = []
	for i in range(len(vertexes)):
		if not random.choices([False, True], [fail_prob[i], 1 - fail_prob[i]])[0]:
			failed_items.append(vertexes[i])
	return failed_items

#### *a)*
Với *Câu 1a*, ta điều chỉnh mảng ` fail_prob` để xác suất hỏng của $R_1$ là $1$, đảm bảo $R_1$ luôn được thêm vào `failed_items`.

*Kết quả tính toán:* $0.8075$  
*Kết quả mô phỏng*:

In [19]:
n = 1000000 # The smaller the faster 

def number_of_successes(n):
    k = 0
    for _ in range(n):
        failed_items = make_fail_items()
        if find_path('A', [], 'B', failed_items):
            k += 1
    return k

# Câu 1a
fail_prob = [0, 1, 0.15, 0.2, 0.25, 0] # P("R1 fails") = 1
print("Câu 1a: ", number_of_successes(n)/n)

Câu 1a:  0.806972


#### *b)*  
*Kết quả tính toán:* $0.98075$  
*Kết quả mô phỏng*:

In [3]:
n = 100000
# Câu 1b
fail_prob = [0, 0.1, 0.15, 0.2, 0.25, 0]
print("Câu 1b: ", number_of_successes(n)/n)

Câu 1b:  0.980875


#### **Câu 2.**
#### *a)*  
Ma trận chuyển trạng thái của chuột là
$P =
\begin{bmatrix}
p_1 & p_2 & p_3 & p_4
\end{bmatrix}
$ $=$
$
\begin{bmatrix}
1/2 & 1/4 & 0 & 0 \\
1/2 & 1/4 & 1/3 & 1/3\\
0 & 1/4 & 1/3 & 1/3\\
0 & 1/4 & 1/3 & 1/3
\end{bmatrix}
$  

Ban đầu phân phối xác suất là:
$$\pi_0 = \begin{bmatrix}
1 & 0 & 0 & 0
\end{bmatrix}$$

Phân phối xác suất tại thời điểm $t = 5$ ngày sau là:
$$\pi_5 = P\times \pi_0 = {\begin{bmatrix}
1/2 & 1/4 & 0 & 0 \\
1/2 & 1/4 & 1/3 & 1/3\\
0 & 1/4 & 1/3 & 1/3\\
0 & 1/4 & 1/3 & 1/3
\end{bmatrix}}^5 \times {\begin{bmatrix}
1 & 0 & 0 & 0
\end{bmatrix}
} $$


In [43]:
import numpy as np
p_0 = np.array([1, 0, 0, 0])
P = np.array([[1/2, 1/4, 0, 0],
              [1/2, 1/4, 1/3, 1/3],
              [0, 1/4, 1/3, 1/3],
              [0, 1/4, 1/3, 1/3]])
p_5 = np.dot(np.linalg.matrix_power(P, 5), p_0)
print(p_5)

[0.20334201 0.34280961 0.22692419 0.22692419]
1.0



Vậy:  
$$\pi_5 = 
\begin{bmatrix}
0.20334201 & 0.34280961 & 0.22692419 & 0.22692419
\end{bmatrix}$$

Vậy xác suất chuột vẫn ở phòng $1$ nếu ban đầu ở phòng $1$ là $p_{1,1} = 0.203342$

#### *b)*  
Khi $n$ tiến ra $\infty$, $n-1$ cũng tiến ra $\infty$.  
Vậy khi đó:  
$$\pi_n = P\times \pi_n$$

Tức là:  
$$(I_n - P)\pi_n = 0$$

Với $\pi_n = \begin{bmatrix}
a, b, c, d
\end{bmatrix}$ và $(I_n - P) = \begin{bmatrix}
1/2 & -1/4 & 0 & 0 \\
-1/2 & 3/4 & -1/3 & -1/3\\
0 & -1/4 & 2/3 & -1/3\\
0 & -1/4 & -1/3 & 2/3
\end{bmatrix}$

Giải hệ phương trình tuyến tính $4$ ẩn trên với điều kiện $a + b + c + d = 1$:


In [50]:
i_p = np.array([[1/2, -1/4, 0, 0],
              [-1/2, 3/4, -1/3, -1/3],
              [0, -1/4, 2/3, -1/3],
              [1, 1, 1, 1]])
p_n = np.array([0, 0, 0, 1])
print(np.linalg.solve(i_p, p_n))

[0.16666667 0.33333333 0.25       0.25      ]


Suy ra ta được: 
\begin{cases}
a = \frac{1}{6} = 0.1667 \\
b = \frac{1}{3} = 0.3333 \\
c = \frac{1}{4} = 0.25 \\
d = \frac{1}{4} = 0.25
\end{cases}

Vậy sau $n$ ngày thì xác suất chuột di chuyển sang các phòng khác là:  
$$\pi_n =  \begin{bmatrix}
0.1667 & 0.3333 & 0.25 & 0.25
\end{bmatrix}$$

#### *c)*  
Gọi $p_{(a, b)}$ là vectơ xác suất của chuột khi ở phòng $a$ được $b$ ngày ($a \in \overline{1, 4}, \;\; b \in \overline{1, 3}$). Ma trận chuyển trạng thái của chuột bây giờ là  
$$
\begin{align*}
    P &=
    \begin{bmatrix}
    p_{(1,1)} & p_{(1,2)} & p_{(1,3)} & p_{(2,1)} & p_{(2,2)} & p_{(2,3)} & p_{(3,1)} & p_{(3,2)} & p_{(3,3)} & p_{(4,1)} & p_{(4,2)} & p_{(4,3)} 
    \end{bmatrix} \\[10pt]
    &=
    \begin{bmatrix}
    0   & 0   & 0     & 1/4 & 1/4 & 1/3    & 0   & 0   & 0       & 0   & 0   & 0 \\
    1/2 & 0   & 0     & 0   & 0   & 0      & 0   & 0   & 0       & 0   & 0   & 0 \\
    0   & 1/2 & 0     & 0   & 0   & 0      & 0   & 0   & 0       & 0   & 0   & 0 \\
           
    1/2 & 1/2 & 1     & 0   & 0   & 0      & 1/3 & 1/3 & 1/2     & 1/3 & 1/3 & 1/2 \\
    0   & 0   & 0     & 1/4 & 0   & 0      & 0   & 0   & 0       & 0   & 0   & 0 \\
    0   & 0   & 0     & 0   & 1/4 & 0      & 0   & 0   & 0       & 0   & 0   & 0 \\
               
    0   & 0   & 0     & 1/4 & 1/4 & 1/3    & 0   & 0   & 0       & 1/3 & 1/3 & 1/2  \\
    0   & 0   & 0     & 0   & 0   & 0      & 1/3 & 0   & 0       & 0   & 0   & 0  \\
    0   & 0   & 0     & 0   & 0   & 0      & 0   & 1/3 & 0       & 0   & 0   & 0  \\
               
    0   & 0   & 0     & 1/4 & 1/4 & 1/3    & 1/3 & 1/3 & 1/2     & 0   & 0   & 0 \\
    0   & 0   & 0     & 0   & 0   & 0      & 0   & 0   & 0       & 1/3 & 0   & 0 \\
    0   & 0   & 0     & 0   & 0   & 0      & 0   & 0   & 0       & 0   & 1/3 & 0
    \end{bmatrix}
\end{align*}
$$
Ban đầu phân phối xác suất là:
$$\pi_0 = \begin{bmatrix}
1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
\end{bmatrix}$$

Phân phối xác suất tại thời điểm $t = 5$ ngày sau là:
$$\pi_5 = P \times \pi_0

In [52]:
import numpy as np
p_0 = np.array([1, 0, 0,    0  ,   0,   0,    0  ,   0,   0,     0  ,   0,   0])

P = np.array([[0  ,   0, 0,    1/4, 1/4, 1/3,    0  ,   0,   0,     0  ,   0,   0],
              [1/2,   0, 0,    0  ,   0,   0,    0  ,   0,   0,     0  ,   0,  0],
              [0  , 1/2, 0,    0  ,   0,   0,    0  ,   0,   0,     0  ,   0,   0],
              [1/2, 1/2, 1,    0  ,   0,   0,    1/3, 1/3, 1/2,     1/3, 1/3, 1/2],
              [0  ,   0, 0,    1/4,   0,   0,    0  ,   0,   0,     0  ,   0,   0],
              [0  ,   0, 0,    0  , 1/4,   0,    0  ,   0,   0,     0  ,   0,   0],
              [0  ,   0, 0,    1/4, 1/4, 1/3,    0  ,   0,   0,     1/3, 1/3, 1/2],
              [0  ,   0, 0,    0  ,   0,   0,    1/3,   0,   0,     0  ,   0,   0],
              [0  ,   0, 0,    0  ,   0,   0,    0  , 1/3,   0,     0  ,   0,   0],
              [0  ,   0, 0,    1/4, 1/4, 1/3,    1/3, 1/3, 1/2,     0  ,   0,   0],
              [0  ,   0, 0,    0  ,   0,   0,    0  ,   0,   0,     1/3,   0,   0],
              [0  ,   0, 0,    0  ,   0,   0,    0  ,   0,   0,     0  , 1/3,   0]])

p_5 = np.dot(np.linalg.matrix_power(P, 5), p_0)
print(p_5)

[0.07899306 0.0625     0.0234375  0.28385417 0.04904514 0.02473958
 0.16232639 0.06134259 0.0150463  0.16232639 0.06134259 0.0150463 ]


Vậy:
$$\pi_5 =  \begin{bmatrix}
0.07899306 & 0.0625 & 0.0234375\dots
\end{bmatrix}$$

Tổng xác suất chuột ở phòng $1$ (đã được $1$ ngày, $2$ ngày, $3$ ngày) sau $5$ ngày là:  
$$P(C) = 0.07899306 + 0.0625 + 0.0234375 =  0.16493056$$

#### *d)*  

Với $\pi_n = \begin{bmatrix}
a_1, a_2, a_3, b_1, b_2, b_3, c_1, c_2, c_3, d_1, d_2, d_3
\end{bmatrix}$ và :
$$(I_n - P) =     \begin{bmatrix}
    0   & 0   & 0     & -1/4 & -1/4 & -1/3    & 0   & 0   & 0       & 0   & 0   & 0 \\
    -1/2 & 0   & 0     & 0   & 0   & 0      & 0   & 0   & 0       & 0   & 0   & 0 \\
    0   & -1/2 & 0     & 0   & 0   & 0      & 0   & 0   & 0       & 0   & 0   & 0 \\
           
    -1/2 & -1/2 & -1     & 0   & 0   & 0      & -1/3 & -1/3 & -1/2     & -1/3 & -1/3 & -1/2 \\
    0   & 0   & 0     & -1/4 & 0   & 0      & 0   & 0   & 0       & 0   & 0   & 0 \\
    0   & 0   & 0     & 0   & -1/4 & 0      & 0   & 0   & 0       & 0   & 0   & 0 \\
               
    0   & 0   & 0     & -1/4 & -1/4 & -1/3    & 0   & 0   & 0       & -1/3 & -1/3 & -1/2  \\
    0   & 0   & 0     & 0   & 0   & 0      & -1/3 & 0   & 0       & 0   & 0   & 0  \\
    0   & 0   & 0     & 0   & 0   & 0      & 0   & -1/3 & 0       & 0   & 0   & 0  \\
               
    0   & 0   & 0     & -1/4 & -1/4 & -1/3    & -1/3 & -1/3 & -1/2     & 0   & 0   & 0 \\
    0   & 0   & 0     & 0   & 0   & 0      & 0   & 0   & 0       & -1/3 & 0   & 0 \\
    0   & 0   & 0     & 0   & 0   & 0      & 0   & 0   & 0       & 0   & -1/3 & 0
    \end{bmatrix}$$

Giải hệ phương trình tuyến tính $12$ ẩn trên với điều kiện $\sum_{i = 1}^3{a_i + b_i + c_i + d_i} = 1$:


In [69]:
i_p = np.subtract(np.identity(12), P)
i_p = np.delete(i_p, 11, axis = 0)
i_p = np.vstack((i_p, np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])))

p_n = np.array(np.zeros(12))
p_n = np.delete(p_n, -1)
p_n = np.append(p_n, 1)

print(np.linalg.solve(i_p, p_n))

[0.08721987 0.04360993 0.02180497 0.2616596  0.0654149  0.01635373
 0.17443973 0.05814658 0.01938219 0.17443973 0.05814658 0.01938219]


Vậy ta được xác suất chuột ở từng phòng sau một thời gian dài $n$ ngày là:
$$\begin{cases}
    {\pi_{n,1}}'&=& 0.08721987 + 0.04360993 + 0.02180497 = 0.15263477 \\
    {\pi_{n,2}}'&=& 0.2616596  + 0.0654149  + 0.01635373 = 0.34342823\\
    {\pi_{n,3}}'&=& 0.17443973 + 0.05814658 + 0.01938219 = 0.2519685\\
    {\pi_{n,4}}'&=& 0.17443973 + 0.05814658 + 0.01938219 = 0.2519685
\end{cases}$$

Do đó:
$${\pi_n}' = \begin{bmatrix}
0.15263477 & 0.34342823 & 0.2519685 & 0.2519685
\end{bmatrix}$$

#### **Kiểm tra.**  

*Giải thích:* 
* Hàm `choose_room` và `choose_new_room` giúp chọn ngẫu nhiên một phòng (mới) bất kì cho chuột đi khi biết vị trí của chuột ở `from_room`, các hàm này dựa vào mảng `doors` để xem `from_room` có cửa dẫn đến những phòng nào.
* Hàm `final_room_1` trả về vị trí chuột sau `after_days` ngày bằng cách gọi `after_days` lần hàm `choose_room`, với tham số mỗi lần gọi là vị trí phòng chuột đến tại lần kề trước, hàm `final_room_2` cũng tương tự nhưng có thêm tham số `forcing_days` là số ngày tối đa chuột có thể ở lại một phòng trước khi bị buộc phải sang phòng mới, mặc định là $3$ ngày.


In [71]:
import random

rooms = ['1', '2', '3', '4']
doors = {
    '1': ['1', '2'],
    '2': ['1', '2', '3', '4'],
    '3': ['2', '3', '4'],
    '4': ['2', '3' , '4']
}

def choose_room(from_room):
    return random.choices(doors[from_room], k = 1)[0]

def choose_new_room(from_room):
    doors_list = []
    for room in doors[from_room]:
        if (room != from_room):
            doors_list.append(room)
    return random.choices(doors_list, k = 1)[0]

def final_room_1(first_room, after_days):
    current_room = first_room
    for _ in range(after_days):
        current_room = choose_room(current_room)
    return current_room

def final_room_2(first_room, after_days, forcing_days = 3):
    current_room = first_room
    days_not_move = 1
    for _ in range(after_days):
        if (days_not_move == forcing_days):
            next_room = choose_new_room(current_room)
            current_room = next_room
            days_not_move = 1
            continue
        next_room = choose_room(current_room)
        if (current_room == next_room):
            days_not_move += 1
        else:
            days_not_move = 1
        current_room = next_room
    return current_room

#### *a)*  
*Kết quả tính toán:* $0.203342$  
*Kết quả mô phỏng*:

In [75]:
n = 1000000

def success_times_1(n, from_room, to_room, after_days):
    k = 0
    for _ in range(n):
        if to_room  == final_room_1(from_room, after_days):
            k += 1
    return k

# Câu 2a
print("Câu 2a: ", success_times_1(n, '1', '1', 5)/n)

Câu 2a:  0.203702


#### *b)*  
*Kết quả tính toán:* $\begin{bmatrix}
0.1667 & 0.3333 & 0.25 & 0.25
\end{bmatrix}$  
*Kết quả mô phỏng*:

In [254]:
n = 5000
def rooms_stayed_times_1(n, from_room, after_days):
    k = {'1': 0, '2': 0, '3': 0, '4': 0}
    for _ in range(n):
        k[final_room_1(from_room, after_days)] += 1           
    return k

# Câu 2b
answer_2b =  rooms_stayed_times_1(n, '1', n)
list_2b = [answer_2b['1']/n, answer_2b['2']/n, answer_2b['3']/n, answer_2b['4']/n]
print("Câu 2b: ", list_2b)

Câu 2b:  [0.1702, 0.3392, 0.249, 0.2416]


#### *c)*  
*Kết quả tính toán:* $0.16493056$  
*Kết quả mô phỏng*:

In [255]:
n = 100000

def success_times_2(n, from_room, to_room, after_days):
    k = 0
    for _ in range(n):
        if to_room  == final_room_2(from_room, after_days, 3):
            k += 1
    return k

# Câu 2c
print("Câu 2c: ", success_times_2(n, '1', '1', 5)/n)


Câu 2c:  0.16468


#### *d)*  
*Kết quả tính toán:* $\begin{bmatrix}
0.15263477 & 0.34342823 & 0.2519685 & 0.2519685
\end{bmatrix}$  
*Kết quả mô phỏng*:

In [256]:
n = 5000
def rooms_stayed_times_2(n, from_room, after_days):
    k = {'1': 0, '2': 0, '3': 0, '4': 0}
    for _ in range(n):
        k[final_room_2(from_room, after_days, 3)] += 1           
    return k

# Câu 2d
answer_2d =  rooms_stayed_times_2(n, '1', n)
list_2d = [answer_2d['1']/n, answer_2d['2']/n, answer_2d['3']/n, answer_2d['4']/n]
print("Câu 2d: ", list_2d)

Câu 2d:  [0.1524, 0.3444, 0.2644, 0.2388]


#### **Câu 3.**  

Gọi $T_n$, $L_n$ và $X_n$ lần lượt là số khách trên xe sau khi rời trạm $n$, số người lên tại trạm $n$ và số người xuống tại trạm $n$. Ta có công thức đệ quy sau:
$$\begin{cases}
T_n &= T_{n-1} + L_n - X_n \\
T_{0} &= 0
\end{cases}$$

Giả sử sau khi rời trạm $n$ xe có $k$ người trên xe, gọi $l$, $x$ lần lượt là số người đã lên xe và xuống xe tại trạm $n$. Dễ thấy:  
$$P(T_n = k) = \sum_l \sum_x P(T_{n-1} = k - l + x)\times P(L_n = l) \times P(X_n = x | T_{n - 1} = k - l + x)$$  
Trong đó  
* $P(L_n = l) =$ $\begin{cases}
    0.5 \text{\;\;khi\;\;} l = 0 \\
    0.4 \text{\;\;khi\;\;} l = 1 \\
    0.1 \text{\;\;khi\;\;} l = 2 \\
    0 \text{\;\;khi\;\;} l \notin \{0, 1, 2\} 
\end{cases}$

* $X_n$ là phân phối nhị thức và $P(X_n = x | T_{n - 1} = k - l + x) = C_{k - l + x}^x\times 0.2^x\times 0.8^{k - l}$

**Chú ý:**
* $L_n\leq T_n \leq 2n$
* $0 \leq L_n \leq 2$
* $0 \leq X_n \leq 2(n-1)$

> *Ghi chú:* Với hai câu *a)* và *b)*. Em nghĩ cách dùng từ *"**vẫn** không có khách"* khá nhập nhằng, có thể hiểu là xe chưa đón được khách nào lên từ lúc khởi hành. Xác suất này khá dễ tính, vậy nên em xin hiểu theo cách sau: *Xác suất xe rời trạm $2$  (trạm $3$ đối với ý b)) mà trên xe ~~vẫn~~ không có khách.*  

#### *a)*  

Theo đề bài, ta cần tìm $P(T_2 = 0)$, tức xác suất $T_1 - X_2 + L_2 = 0$. Ta có: 
\begin{align*}
&& 0 \leq L_2 = l \leq T_2 = 0 &\text{\;\;và\;\;}& 0 \leq X_2 = x \leq 2 \\
&\Rightarrow& l = 0 &\text{\;\;và\;\;}& x \in \{0, 1, 2\} 
\end{align*}

Vậy các bộ sau đây thoả đề bài:
$$\begin{cases}
l = 0,\;\; x = 0, T_1 = 0\\
l = 0,\;\; x = 1, T_1 = 1\\
l = 0,\;\; x = 2, T_1 = 2\\
\end{cases}$$

Ta có:
$$\begin{cases}
P(X_2 = 0 | T_1 = 0) &=& C_{0}^0\times 0.2^0\times 0.8^{0} &=& 1 \\
P(X_2 = 1 | T_1 = 1) &=& C_{1}^1\times 0.2^1\times 0.8^{0} &=& 0.2 \\
P(X_2 = 2 | T_1 = 2) &=& C_{2}^2\times 0.2^2\times 0.8^{0} &=& 0.04 \\
\end{cases}
\text{\;\;và\;\;}
P(L_2 = 0) = 0.5
\text{\;\;và\;\;}
P(T_1 = t) = P(L_1 = t) = \begin{cases}
    0.5 \text{\;\;khi\;\;} t = 0 \\
    0.4 \text{\;\;khi\;\;} t = 1 \\
    0.1 \text{\;\;khi\;\;} t = 2 \\
    0 \text{\;\;khi\;\;} t \notin \{0, 1, 2\} 
\end{cases}
$$

Vậy xác suất cần tính là:
$$\begin{align*}
&P(T_2 = 0) &=&\; \sum_l \sum_x P(T_1 = - l + x)\times P(L_2 = 0) \times P(X_2 = x | T_1 = -l + x) \\
&&=&\;\; 0.5\times 0.5\times 1 + 0.4\times 0.5 \times 0.2 + 0.1\times 0.5 \times 0.04 = 0.292
\end{align*}$$


#### *b)*  

Theo đề bài, ta cần tìm $P(T_3 = 0)$, tức xác suất $T_2 - X_3 + L_3 = 0$. Ta có: 
\begin{align*}
&& 0 \leq L_3 = l_3 \leq T_3 = 0 &\text{\;\;và\;\;}& 0 \leq X_3 = x_3 \leq 4 \\
&\Rightarrow& l_3 = 0 &\text{\;\;và\;\;}& x_3 \in \{0, 1, 2, 3, 4\} 
\end{align*}

Khó để giải bài toán này theo cách *a)* trên do có quá nhiều trường hợp, tuy nhiên có thể lập trình bài toán tính $P(T_n = k)$ bằng cách cài đặt đệ quy câu *a)* như sau:
* Hàm `prob_L`, `prob_X` được cài đặt theo công thức đã đề cập.
* Hàm `prob_T` là hàm đệ quy tính giá trị xác suất có `n` trên xe sau khi xe rời trạm `at_station`, với điều kiện dừng tại $P(T_0 = x)$

In [129]:
from numpy import math
import itertools

def prob_L(n):
    if n == 0:
        return 0.5
    if n == 1:
        return 0.4
    if n  == 2:
        return 0.1
    return 0

def prob_X(n, current_size):
    return math.comb(current_size, n) * (0.2 ** n) * (0.8 ** (current_size - n))

def prob_T(n, at_station):
    if at_station == 0:
        return 1 if n == 0 else 0
    list_L = [i for i in range(min(n + 1, 3))] # L_n
    list_X = [i for i in range(2 * ( at_station - 1) + 1)] # X_n
    list_T = [i for i in range(2 * (at_station - 1) + 1)] # T_{n-1}
    total_prob = 0
    for l, x, t in itertools.product(list_L, list_X, list_T):
        if l - x + t != n:
            continue
        if x > t:
            continue
        total_prob += prob_T(t, at_station - 1) * prob_L(l) * prob_X(x, t)
    return total_prob

In [136]:
print('Câu 3b:' ,prob_T(0, 3))

Câu 3b: 0.19183232


Vậy kết quả là: 
$$P(T_3 = 0) = 0.19183232$$

#### *c)*  
Ta đã biết:
$$0 \leq T_5 \leq 2\times 5 = 10$$

Vậy theo đề bài, ta được tập các giá trị của $T_5$ là $\{5, 6, 7, 8, 9, 10\}$. Cài đặt python như sau:

In [139]:
list_T5 = [5, 6, 7, 8, 9, 10]
calc_result_3c = 0
for t in list_T5:
    calc_result_3c += prob_T(t, 5)
print('Câu 3b:' , calc_result_3c)

Câu 3b: 0.03430457666897152


Vậy kết quả là: 
$$P(T_5 \geq 5) = 0.03430457666897152$$

#### *d)*  
Theo đề bài, ta cần tính xác suất:  
$$P(T_1 = \lvert\{An\}\rvert\;|\; T_2 = 0)$$

Mà ta có:
$$P(T_1 = \lvert\{An\}\rvert) = P(T_1 = 1) = L_1 = 0.4$$

Dựa vào câu *a)*, ta cũng có:
$$P(T_2 = 0 \wedge T_1 = \lvert\{An\}\rvert) = P(T_1 = 1) \times P(L_2 = 0) \times P(X_2 = 1 | T_1 = 1) = 0.4\times 0.5 \times 0.2 = 0.04$$

Vậy:
$$P(T_1 = \lvert\{An\}\rvert\;|\; T_2 = 0) = \frac{P(T_2 = 0 \wedge T_1 = \lvert\{An\}\rvert)}{P(T_2 = 0)} = \frac{0.04}{0.292} = 0.13698630137$$

#### *e)*  
Ta biết $0 \leq T_2 \leq 2\times 2 = 4$, vậy tập các giá trị có thể có của $T_2$ là \{0, 1, 2, 3, 4\}. Do đó cài đặt python như sau:

In [147]:
list_t2 = [0, 1, 2, 3, 4]
result_3e = 0
for t in list_t2:
    result_3e += t * prob_T(t, 2)
print("Câu 3e: ", result_3e)

Câu 3e:  1.0800000000000003


Vậy ta có:
$$E(T_2) = 1.08$$

#### *f)*  
Tương tự, ta có tập giá trị của $T_3$ là $\{0, 1, 2, 3, 4, 5, 6\}$.

In [150]:
list_t3 = [0, 1, 2, 3, 4, 5, 6]
result_3f = 0
for t in list_t3:
    result_3f += t * prob_T(t, 3)
print("Câu 3f: ", result_3f)

Câu 3f:  1.4640000000000006


Vậy ta có:
$$E(T_3) = 1.464$$

#### *g)*
Nếu số người lên cũng phụ thuộc vào số chỗ còn lại trên xe, giờ đây $P(L_n = l)$ cũng bị phụ thuộc vào $P(T_{n-1} = t)$. Ta phải đảm bảo rằng:  
$$T_{n-1} - X_n + L_n \leq 5 \text{\;\;với\;\;}T_{n-1}-X_{n} = \Delta \leq T_{n-1} \leq 5$$  
* Trường hợp $\Delta \leq 3$:  
Ta có xác suất $P(L_n = l)$ như cũ và được tính như sau:
$P(L_n = l | T_{n - 1} - X_n = \Delta) =$ $\begin{cases}
    0.5 \text{\;\;khi\;\;} l = 0 \\
    0.4 \text{\;\;khi\;\;} l = 1 \\
    0.1 \text{\;\;khi\;\;} l = 2 \\
    0 \text{\;\;khi\;\;} l \notin \{0, 1, 2\} 
\end{cases}$  
* Trường hợp $\Delta = 4$:  
Ta có 
$P(L_n = l | T_{n - 1} - X_n = \Delta) =$ $\begin{cases}
    0.5 \text{\;\;khi\;\;} l = 0 \\
    0.5 \text{\;\;khi\;\;} l = 1 \\
    0 \text{\;\;khi\;\;} l \notin \{0, 1\} 
\end{cases}$  
* Trường hợp $\Delta = 5$:  
Ta có 
$P(L_n = l | T_{n - 1} - X_n = \Delta) =$ $\begin{cases}
    1 \text{\;\;khi\;\;} l = 0 \\
    0 \text{\;\;khi\;\;} l \neq 0
\end{cases}$  
Vậy, cần cài đặt lại hàm `prob_L_2` sao cho nhận thêm tham số `delta`.
Ta cũng cần phải điều chỉnh hàm `prob_T` để loại bỏ bất kì thời điểm nào $T_n > 5$. Các hàm mới như sau:

In [240]:
def prob_L_2(n, delta):
    if delta <= 3:
        return prob_L(n)
    if delta == 4:
        if n == 0:
            return 0.5
        if n == 1:
            return 0.5
        return 0
    if n == 0:
        return 1
    return 0

def prob_T_2(n, at_station):
    if n > 5:
        return 0
    if at_station == 0:
        return 1 if n == 0 else 0
    list_L = [i for i in range(min(n + 1, 3))] # L_n
    list_X = [i for i in range(min(2 * (at_station - 1) + 1,6))] # X_n
    list_T = [i for i in range(min(2 * (at_station - 1) + 1,6))] # T_{n-1}
    total_prob = 0
    for l, x, t in itertools.product(list_L, list_X, list_T):
        if l - x + t != n:
            continue
        if x > t:
            continue
        total_prob += prob_T_2(t, at_station - 1) * prob_L_2(l, t - x) * prob_X(x, t)
    return total_prob

In [241]:
print('Câu 3g:' , prob_T_2(5, 5))

Câu 3g: 0.033708699289649195


Vậy ta có kết quả:
$$P'(T_5 = 5) = 0.033708699289649195$$

#### *h)*
Tương tự như câu *f)*, nhưng với điều kiện chỗ ngồi tối đa, ta có tập giá trị của $T_5$ là $\{0, 1, 2, 3, 4, 5\}$. Cài đặt python là:

In [242]:
list_t5_2 = [0, 1, 2, 3, 4, 5]
result_3h = 0
for t in list_t5_2:
    result_3h += t * prob_T_2(t, 5)
print("Câu 3h: ", result_3h)

Câu 3h:  2.0082514759868815


Vậy ta có kết quả:
$$E'(T_5) = 2.0082514759868815$$

#### **Kiểm tra.**  

*Giải thích:* 
* Hàm `go_down` và `go_up` tính số lượng khách xuống và lên mỗi lượt trạm.
* Hàm `bus_size` tính số lượng hành khách trên xe sau khi xe rời trạm `after_stops`. Tương tự, hàm `bus_size_max` cài đặt thêm tính năng chỉ cho vừa đủ người lên xe nếu số lượng hành khách sắp lên vượt qua `max`.


In [107]:
go_down_probability = 0.2
go_up_numbers = [0, 1, 2]
go_up_probabilty = [0.5, 0.4, 0.1]

def go_down(current_size):
    total_go_down = 0
    for _ in range(current_size):
        total_go_down += random.choices([1, 0], [go_down_probability, 1 - go_down_probability])[0]
    return total_go_down

def go_up():
    return random.choices(go_up_numbers, go_up_probabilty)[0]

def bus_size(after_stops):
    current_size = 0
    for _ in range(after_stops):
        current_size -= go_down(current_size)
        current_size += go_up()
    return current_size

def bus_size_max(after_stops, max):
    current_size = 0
    go_up_people = 0
    for _ in range(after_stops):
        current_size -= go_down(current_size)
        go_up_people = go_up()
        if current_size + go_up_people > max:
            current_size = max
        else:
            current_size += go_up_people
    return current_size

#### *a)*  
*Kết quả tính toán:* $0.292$   
*Kết quả mô phỏng*:

In [244]:
n = 1000000
# Câu 3a
result_3a = 0
for _ in range(n):
    if bus_size(2) == 0:
        result_3a += 1
print("Câu 3a: ", result_3a/n)

Câu 3a:  0.29142


#### *b)*  
*Kết quả tính toán:* $0.19183232$   
*Kết quả mô phỏng*:

In [245]:
n = 1000000
# Câu 3b
result_3b = 0
for _ in range(n):
    if bus_size(3) == 0:
        result_3b += 1
print("Câu 3b: ", result_3b/n)

Câu 3b:  0.191776


#### *c)*  
*Kết quả tính toán:* $0.03430457666897152$   
*Kết quả mô phỏng*:

In [246]:
n = 1000000
# Câu 3c
result_3c = 0
for _ in range(n):
    if bus_size(5) >= 5:
        result_3c += 1
print("Câu 3c: ", result_3c/n)

Câu 3c:  0.034173


#### *d)*  
*Kết quả tính toán:* $0.13698630137$   
*Kết quả mô phỏng*:

In [247]:
n = 1000000
# Câu 3d
result_3d = 0
m = 0
for _ in range(n):
    t1 = bus_size(1)
    t2 = t1
    t2 -= go_down(t2)
    t2 += go_up()

    if t2 == 0:
        m += 1
        if t1 == 1:
            result_3d += 1

print("Câu 3d: ", result_3d/m)

Câu 3d:  0.1369052318656007


#### *e)*  
*Kết quả tính toán:* $1.08$   
*Kết quả mô phỏng*:

In [248]:
n = 1000000
# Câu 3e
result_3e = 0
for _ in range(n):
    result_3e += bus_size(2)
print("Câu 3e: ", result_3e/n)

Câu 3e:  1.079489


#### *f)*  
*Kết quả tính toán:* $1.464$   
*Kết quả mô phỏng*:

In [249]:
n = 1000000
# Câu 3f
result_3f = 0
for _ in range(n):
    result_3f += bus_size(3)
print("Câu 3f: ", result_3f/n)

Câu 3f:  1.464782


#### *g)*  
*Kết quả tính toán:* $0.033708699289649195$   
*Kết quả mô phỏng*:

In [250]:
n = 1000000
# Câu 3g
result_3g = 0
for i in range(n):
    if bus_size_max(5, 5) == 5:
        result_3g += 1
print("Câu 3g: ", result_3g/n)

Câu 3g:  0.033531


#### *h)*  
*Kết quả tính toán:* $2.0082514759868815$   
*Kết quả mô phỏng*:

In [252]:
n = 1000000
# Câu 3h
result_3h = 0
for _ in range(n):
    result_3h += bus_size_max(5, 5)
print("Câu 3h: ", result_3h/n)

Câu 3h:  2.009115
